src/core/unif.ml

```diff
@@ -76,29 +76,30 @@ let instantiate : meta -> term array -> term -> bool = fun m ts u ->
         && (set_meta m (Bindlib.unbox bu); true)
 
 (** [solve cfg p] tries to solve the unification problems of [p] and
-   returns the constraints that could not be solved. *)
+    returns the constraints that could not be solved. *)
 let rec solve : problems -> unif_constrs = fun p ->
-  (*if !log_enabled then log_unif "%a" pp_problem p;*)
   match p with
   | { to_solve = []; unsolved = []; _ } -> []
   | { to_solve = []; unsolved = cs; recompute = true } ->
      solve {no_problems with to_solve = cs}
   | { to_solve = []; unsolved = cs; _ } -> cs
-  | { to_solve = (t,u)::to_solve; _ } -> solve_aux t u {p with to_solve}
+  | { to_solve = (c,t,u)::to_solve; _ } -> solve_aux c t u {p with to_solve}
 
 (** [solve_aux t1 t2 p] tries to solve the unificaton problem given by [p] and
     the constraint [(t1,t2)], starting with the latter. *)
-and solve_aux : term -> term -> problems -> unif_constrs = fun t1 t2 p ->
+and solve_aux : ctxt -> term -> term -> problems -> unif_constrs =
+  fun ctx t1 t2 p ->
   let (h1, ts1) = Eval.whnf_stk t1 [] in
```

Comment:
There is a problem here because ctx (which may contain defined variables) is not taken into account in Eval.whnf.

Reply:
The function Eval.whnf should take a context as argument.

Reply:
To do so, you could define a function Ctxt.to_let : ctxt -> term -> term which adds and binds on top of the term argument a let for each defined variable of the ctxt argument.

Reply:
So I guess if the variable is at the head of the term, it should be unfolded

Reply:
But wouldn't this call to `to_let` be redundant with the fact that the `solve` functions unfold the variables? (in the big pattern matching at the end)

Reply:
If we modify whnf then, indeed, variable unfolding can be removed in solve_aux.
I think that this is better to preserve the meaning of what is a whnf.

Reply:
But in that case, it becomes a whnf relatively to a context, which changes the meaning anyway. Do we need this kind of whnf somewhere else? If not, I think it would be better to leave context management to the code of `unif.ml`.

Reply:
It is important to preserve the meaning of whnf. The use of contexts with defined variables is not restricted to unification. Such contexts will also appear in proofs (we need to add a tactic for that though but we will do it soon or later as it is a very useful one). Anyway, infer is defined on contexts with defined variables, and infer needs to compute whnf. If defined variables are not taken into account in the computation of whnf, then the type inference algorithm is incomplete.

Reply:
I see.

Reply:
> To do so, you could define a function Ctxt.to_let : ctxt -> term -> term which adds and binds on top of the term argument a let for each defined variable of the ctxt argument.

It may be inefficient to build the let-bindings to remove them just after

Reply:
Perhaps but, in general, contexts do not contain many defined variables. And we may try to improve that later.

Reply:
Btw, note that @barras is working on a more efficient reduction machine, and will provide some PR soon.

Reply:
OK. Are you happy with what I have done? (without the conversion to lets)

Reply:
Because of https://github.com/Deducteam/lambdapi/pull/309#issuecomment-603964495, no. See https://github.com/Deducteam/lambdapi/pull/309#issuecomment-603977408 to fix it I think.